### DISCUSSION TITLE
synchronous execution in a `tokio::spawn`

### DISCUSSION THREAD

**Discussion by setoelkahfi** [tokio-rs/tokio#6857] [Q&A] [answered]
I spawned a new thread with `tokio::spawn` and some asynchronous tasks. Like so:

```rust
    fn prepare_result_files(app: tauri::AppHandle, audio_file_id: i32, results: Vec<ResultFile>) {
        debug!("Preparing result files: {:?}", audio_file_id);
        let id = audio_file_id.to_string();
        tokio::spawn(async move {
            save_result_files(&id, &results).await;
            download_result_files(&id, results).await;
            prepare_tracks();
            prepare_players(&id).await;
            let _ = app.emit_all(PLAYER_IS_READY, audio_file_id);
        });
    }
```

The problem is that the `emil_all` event sometimes gets executed before all the `await` tasks are finished. I want to perform the functions inside the spawn synchronously so that the code emits `PLAYER_IS_READY` event correctly. I have looked at the `spawn_blocking` one, but I don't think that's what I need.

Any pointers?

SO [thread](https://stackoverflow.com/questions/79011210/synchronous-execution-in-a-tokiospawn).

**Comment by Darksonn** [comment, marked as answer]
Async/await executes operations one line at the time, so `emit_all` does not get called before the `.await` operations before it finish.

**Comment by setoelkahfi** [comment]
Ok, I think this is a bug somewhere else. Thanks for the reply @Darksonn
